python_modules/dagster/dagster/_core/definitions/partition.py

```diff
@@ -732,13 +750,20 @@ def get_run_config_for_partition_key(
         partition_key: str,
         dynamic_partitions_store: Optional[DynamicPartitionsStore] = None,
         current_time: Optional[datetime] = None,
+        check_valid_partition_key=True,
```

Comment:
I don't love this pattern...

The alternative is making the `get_run_config_for_partition_key` and `get_tags_for_partition_key` accept a partition object. Then we'd create a `PendingPartition` to pass into those functions. I think this is more performant since we won't be refetch the same partition object multiple times (as we are now), but I don't want to enable this if we plan on removing `Partition` in the near future.

Reply:
Oh wow, I forgot we were doing this O(N) operation every time `get_run_config_for_partition_key` is called.

I don't love this pattern either.  I think it would be cool to remove or at least sequester `Partition`, but probably not best to wait on that for this PR.

A couple different ideas here:
- Add a boolean property to `PartitionsDefinition` that lets implementations declare whether there's any difference between the values inside their `Partition` objects and their `partition_key`. It would be false for everything except for `TimeWindowPartitionsDefinition` I think? Although what does the `Partition` look like for a `MultiPartitionsDefinition` that contains a `TimeWindowPartitionsDefinition`?
- Add a method to `PartitionsDefinition` that accepts a `partition_key` and returns a `Partition` object.

Reply:
before 1.2.4 we were not--this function just accepted a `Partition` object, which ensured that the partition was valid. But after the removal of `PartitionSet`, these functions now check for the existence of the partition within each call.

> Add a method to PartitionsDefinition that accepts a partition_key and returns a Partition object.

This option sounds good to me, since we can ensure that the partition exists before creating the Partition object. In the future when we remove `Partition`, maybe we can create some internal `ValidatedPartitionKey` object that we pass around to these methods to avoid revalidating on each function call.